llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp

```diff
@@ -104,6 +109,24 @@ bool SPIRVEmitNonSemanticDI::emitGlobalDI(MachineFunction &MF) {
                 cast<ConstantAsMetadata>(Op->getOperand(2))->getValue())
                 ->getSExtValue();
     }
+
+    // This traversal is the only supported way to access
+    // instruction related DI metadata like DIBasicType
+    for (auto &F : *M) {
+      for (auto &BB : F) {
```

Comment:
Could iteration over BBs be skipped? Please see this _[Iterating over the Instruction in a Function section in the LLVM Programmers Manual](https://llvm.org/docs/ProgrammersManual.html#iterating-over-the-instruction-in-a-function)_.

Reply:
```
    for (const Function &F : *M) {
      for (inst_iterator I = inst_begin(&F), E = inst_end(&F); I != E; ++I) {
        for (DbgRecord &DR : I->getDbgRecordRange()) {
```
can't make it work on current trunk:
```
llvm/include/llvm/IR/InstIterator.h:54:14: error: 'BBs' is a private member of 'llvm::InstIterator<const llvm::SymbolTableList<llvm::BasicBlock>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::BasicBlock, true, false, void, false, void>, false, true>, llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, true>, const llvm::Instruction>'
   54 |     : BBs(II.BBs), BB(II.BB), BI(II.BI) {}
```
I think that error means that during construction of `inst_iterator` some parameters of the template `llvm::ilist_iterator` have private scope on required members. Seems like internal API problem.

Reply:
Please see also available API in https://llvm.org/docs/RemoveDIsDebugInfo.html, https://github.com/llvm/llvm-project/blob/daf208598b06a0b8b3dc3c30b023334182bdf170/llvm/include/llvm/IR/DebugProgramInstruction.h#L547 and an example of usage in https://github.com/llvm/llvm-project/blob/daf208598b06a0b8b3dc3c30b023334182bdf170/llvm/lib/Bitcode/Reader/MetadataLoader.cpp#L623

Reply:
Does this API work for you?

Reply:
The one you provided yes. It's in the current version.